src/main/scala/firrtl/transforms/Dedup.scala

```diff
@@ -18,134 +19,301 @@ case class NoDedupAnnotation(target: ModuleName) extends SingleTargetAnnotation[
+      case d: DefMemory =>
+        val oldType = MemPortUtils.memType(d)
+        val newType = retype(d.name)(oldType)
+        val index = oldType
+          .asInstanceOf[BundleType].fields.headOption
+          .map(_.tpe.asInstanceOf[BundleType].fields.indexWhere(
+            {
+              case Field("data", _, _) => true
+              case Field("wdata", _, _) => true
+              case Field("rdata", _, _) => true
+              case _ => false
+            }))
+        val newDataType = if(index.nonEmpty) {
+          //If index nonempty, then there exists a port
+          val i = index.get
+          newType.asInstanceOf[BundleType].fields.head.tpe.asInstanceOf[BundleType].fields(i).tpe
+        } else {
+          //If index is empty, this mem has no ports, and so we don't need to record the dataType
+          retype(d.name + ";&*^$")(d.dataType)
+        }
+        d.copy(dataType = newDataType) map rename map reinfo
+      case h: IsDeclaration => h map rename map retype(h.name) map onExp map reinfo
+      case other => other map reinfo map onExp map onStmt
     }
-
-    while ((moduleOrder.size < c.modules.size)) {
-      c.modules.foreach(m => if (!moduleOrder.contains(m.name)) addModule(m))
+    val finalModule = module match {
+      case m: Module => m map onPort map onStmt
+      case other => other
     }
-    moduleOrder
+    finalModule
   }
 
-  // Finds duplicate Modules
-  // Also changes DefInstances to instantiate the deduplicated module
-  // Returns (Deduped Module name -> Seq of identical modules,
-  //          Deuplicate Module name -> deduped module name)
-  private def findDups(
-      moduleOrder: Seq[String],
-      moduleMap: Map[String, DefModule],
-      noDedups: Seq[String]): (Map[String, Seq[DefModule]], Map[String, String]) = {
-    // Module body -> Module name
-    val dedupModules = mutable.HashMap.empty[String, String]
-    // Old module name -> dup module name
-    val dedupMap = mutable.HashMap.empty[String, String]
-    // Deduplicated module name -> all identical modules
-    val oldModuleMap = mutable.HashMap.empty[String, Seq[DefModule]]
-
-    def onModule(m: DefModule): Unit = {
-      def fixInstance(s: Statement): Statement = s map fixInstance match {
-        case DefInstance(i, n, m) => DefInstance(i, n, dedupMap.getOrElse(m, m))
-        case WDefInstance(i, n, m, t) => WDefInstance(i, n, dedupMap.getOrElse(m, m), t)
-        case x => x
+  /**
+    * Turns a module into a name-agnostic module
+    * @param module module to change
+    * @return name-agnostic module
+    */
+  def agnostify(module: DefModule, name2tag: mutable.HashMap[String, String], tag2name: mutable.HashMap[String, String]): DefModule = {
+    val namespace = Namespace()
+    val nameMap = mutable.HashMap[String, String]()
+    val typeMap = mutable.HashMap[String, Type]()
+    def rename(name: String): String = {
+      if(nameMap.contains(name)) nameMap(name) else {
```

Comment:
This is a minor thing but canonical Scala style calls for a space after `if`, eg. `if (nameMap.contains(name))...`